File: PetJourneyBackend/app/schemas/web/transport.py
"""平行交通契约：现实参考与动物世界身份分离、门到门 JourneyLeg、地图车辆与活动入口。

时间一律保存完整 UTC；当地展示使用节点时区（IANA）。计划/预计/实际三组时间
分别可缺省——schema 不把实际时间默认为计划时间，也不把估算标成已核验。
"""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum

from pydantic import Field

from .common import DataOrigin, LatLng, WebModel


class TransportMode(str, Enum):
    flight = "flight"
    train = "train"
    ferry = "ferry"
    drive = "drive"
    transit = "transit"
    taxi = "taxi"
    walk = "walk"


class TimeBasis(str, Enum):
    verified_timetable = "verified_timetable"
    live_status = "live_status"
    routed_estimate = "routed_estimate"
    demo_fixture = "demo_fixture"


class LegTimeSource(str, Enum):
    """0.4.0：一段交通时间的精确来源（time_basis 只保留四个大类以兼容旧客户端）。

    verified_timetable 已核验时刻表的班次；operator_rule 运营方规定（例如开船前 30 分钟登船）；routed_estimate 地图供应商的路线估算；
    world_rule 星球内的路程（世界规则估算，不是现实路线；time_basis 记为 routed_estimate＝“估算”）；demo_fixture 演示环境。"""

    verified_timetable = "verified_timetable"
    operator_rule = "operator_rule"
    routed_estimate = "routed_estimate"
    world_rule = "world_rule"
    demo_fixture = "demo_fixture"


class DataFreshness(str, Enum):
    verified = "verified"
    stale = "stale"
    unavailable = "unavailable"


class PositionBasis(str, Enum):
    simulated_route = "simulated_route"
    schematic = "schematic"
    live_vehicle = "live_vehicle"


class TravellerRole(str, Enum):
    driver = "driver"
    passenger = "passenger"
    walker = "walker"


class LegKind(str, Enum):
    main = "main"
    connection = "connection"
    wait = "wait"
    transfer = "transfer"


class LegPhase(str, Enum):
    scheduled = "scheduled"
    connecting = "connecting"
    waiting = "waiting"
    boarding = "boarding"
    in_transit = "in_transit"
    arriving = "arriving"
    arrived = "arrived"
    cancelled = "cancelled"


class NodeKind(str, Enum):
    airport = "airport"
    station = "station"
    port = "port"
    road_point = "road_point"
    place = "place"


class TransportNode(WebModel):
    """verified=False 表示示意节点（例如旧代码插值出的中转点），不得冒充真实枢纽。"""

    node_id: str
    name: str
    kind: NodeKind
    timezone: str = Field(description="IANA 时区，例如 Asia/Hong_Kong")
    lat: float | None = None
    lng: float | None = None
    verified: bool = False


class TransportReference(WebModel):
    """现实参考：仅服务端核验与来源说明使用；主界面展示 WorldService。"""

    reference_id: str
    provider: str
    operating_instance_id: str
    service_date: date
    operator_name: str | None = None
    marketing_codes: list[str] = Field(default_factory=list)
    origin: TransportNode
    destination: TransportNode
    scheduled_departure_utc: datetime | None = None
    scheduled_arrival_utc: datetime | None = None
    estimated_departure_utc: datetime | None = None
    estimated_arrival_utc: datetime | None = None
    actual_departure_utc: datetime | None = None
    actual_arrival_utc: datetime | None = None
    time_basis: TimeBasis
    freshness: DataFreshness
    source_url: str | None = None
    fetched_at: datetime | None = None
    verified_at: datetime | None = None
    usage_scope: str | None = None


class TransportReferenceSummary(WebModel):
    """发给客户端的最小来源说明，不含现实班次号作为主展示。"""

    reference_id: str | None = None
    time_basis: TimeBasis
    freshness: DataFreshness
    source_label: str
    verified_at: datetime | None = None


class WorldService(WebModel):
    """动物世界承运身份：同一运营实例稳定映射（唯一约束 + mapping_version），刷新不换号。"""

    world_service_id: str
    carrier_name: str
    service_code: str
    mode: TransportMode
    vehicle_style: str | None = None
    reference_id: str | None = None
    mapping_version: int = 1


class LegTimes(WebModel):
    origin_timezone: str
    destination_timezone: str
    planned_departure_utc: datetime
    planned_arrival_utc: datetime
    estimated_departure_utc: datetime | None = None
    estimated_arrival_utc: datetime | None = None
    actual_departure_utc: datetime | None = None
    actual_arrival_utc: datetime | None = None


class JourneyLeg(WebModel):
    leg_id: str
    journey_id: str
    sequence: int
    kind: LegKind
    mode: TransportMode
    role: TravellerRole
    world_service: WorldService | None = None
    origin: TransportNode
    destination: TransportNode
    times: LegTimes
    time_basis: TimeBasis
    freshness: DataFreshness
    position_basis: PositionBasis
    phase: LegPhase
    itinerary_version: int
    route: list[LatLng] = Field(default_factory=list, description="有几何才沿线呈现；仅节点时为示意")
    reference: TransportReferenceSummary | None = None
    rescheduled_reason: str | None = None
    time_source: LegTimeSource | None = Field(default=None, description="这段时间的精确来源（见 LegTimeSource）；world_rule 的 time_basis 为 routed_estimate",
                                              json_schema_extra={"x-additive": True})


class VehiclePosition(WebModel):
    """由服务器时间与已确认时间线计算；客户端只做平滑插值呈现，不决定到达。"""

    leg_id: str
    lat: float
    lng: float
    heading_deg: float = 0.0
    progress: float = Field(ge=0.0, le=1.0)
    computed_at: datetime
    position_basis: PositionBasis


class TravelActivityKind(str, Enum):
    listening = "listening"
    watching = "watching"
    resting = "resting"
    window_gazing = "window_gazing"
    dining = "dining"
    writing_postcard = "writing_postcard"


class TravelActivityState(str, Enum):
    active = "active"
    paused = "paused"
    interrupted = "interrupted"
    ended = "ended"


class TravelActivity(WebModel):
    activity_id: str
    leg_id: str
    kind: TravelActivityKind
    state: TravelActivityState
    starts_at: datetime
    ends_at: datetime | None = None
    media_session_id: str | None = None
    interruptible: bool = True
    version: int = 1


class ActivityBadgeKind(str, Enum):
    music = "music"
    tv = "tv"


class MapEntryAction(str, Enum):
    join = "join"
    solo = "solo"
    open_leg_card = "open_leg_card"


class MapActivityEntry(WebModel):
    """地图车辆旁的音符/电视入口；badge 由实际 TravelActivity 决定，无活动则不下发。"""

    entry_id: str
    leg_id: str
    activity_id: str
    badge: ActivityBadgeKind
    badge_state: TravelActivityState
    media_session_id: str | None = None
    label: str
    actions: list[MapEntryAction] = Field(default_factory=list)


class PetArrivalContext(WebModel):
    """Journey → FoodDiscovery 的时间契约：门到门可行到达与可停留窗口，不只看落地时刻。"""

    journey_id: str
    itinerary_version: int
    leg_id: str
    city: str
    destination_timezone: str
    feasible_arrival_utc: datetime
    stay_window_start_utc: datetime
    stay_window_end_utc: datetime


class JourneyMapSnapshot(WebModel):
    journey_id: str
    pet_id: str
    itinerary_version: int
    lifecycle: str = Field(default="active", description="active / completed / cancelled")
    destination_title: str | None = None
    current_visit_id: str | None = Field(default=None, description="宠物此刻在店里时才有值")
    planned_visit_id: str | None = Field(default=None, description="本次旅程的到访（出发即确定；到店前可按寻味推荐改选）")
    server_time: datetime
    legs: list[JourneyLeg] = Field(default_factory=list)
    current_leg_id: str | None = None
    vehicle: VehiclePosition | None = None
    activities: list[TravelActivity] = Field(default_factory=list)
    activity_entries: list[MapActivityEntry] = Field(default_factory=list)
    arrival_context: PetArrivalContext | None = None
    data_origin: DataOrigin
    catching_up: bool = Field(default=False, description="有已经到期、后台还没结算到的事实（到站、收工、回家……）：位置照常按时间显示，工资与来信稍后出现；"
                              "不是出错（0.4.1）", json_schema_extra={"x-additive": True})



class BasemapProvider(str, Enum):
    amap = "amap"


class BasemapUnavailableReason(str, Enum):
    not_configured = "not_configured"
    outside_region = "outside_region"
    daily_cap = "daily_cap"
    user_limit = "user_limit"
    upstream_error = "upstream_error"


class BasemapView(WebModel):
    """旅途地图的真实底图（服务端代理高德静态地图）。available=false 时前端退回示意地图，并说明原因。"""

    available: bool
    reason: BasemapUnavailableReason | None = None
    provider: BasemapProvider | None = None
    image_url: str | None = Field(default=None, description="同源图片地址（需登录），图片为 2 倍清晰度；图上自带供应商标志与审图号，不得遮挡或裁掉")
    center: LatLng | None = Field(default=None, description="图片中心（WGS-84）；叠加物按 WGS-84 做 Web Mercator 投影即可对齐")
    zoom: int | None = Field(default=None, description="Web Mercator 缩放级别（256 像素瓦片）")
    width: int | None = Field(default=None, description="图片逻辑宽度（CSS 像素）")
    height: int | None = Field(default=None, description="图片逻辑高度（CSS 像素）")
    attribution: str | None = None
    expires_at: datetime | None = None

__all__ = [
    "TransportMode",
    "TimeBasis",
    "LegTimeSource",
    "DataFreshness",
    "PositionBasis",
    "TravellerRole",
    "LegKind",
    "LegPhase",
    "NodeKind",
    "TransportNode",
    "TransportReference",
    "TransportReferenceSummary",
    "WorldService",
    "LegTimes",
    "JourneyLeg",
    "VehiclePosition",
    "TravelActivityKind",
    "TravelActivityState",
    "TravelActivity",
    "ActivityBadgeKind",
    "MapEntryAction",
    "MapActivityEntry",
    "PetArrivalContext",
    "JourneyMapSnapshot",
    "BasemapProvider",
    "BasemapUnavailableReason",
    "BasemapView",
]

File: PetJourneyWeb/src/fixtures/transport.ts
/**
 * 平行交通 fixture：飞/铁/船/驾四类 + 暂停/失败/到站变体。
 * 全部 time_basis=demo_fixture：不是已核验时刻表，不对应任何真实班次；承运人名称（喵航 Cat222 等）
 * 是原创动物世界身份示例。起终点坐标仅用于示意地图定位。
 * 时间相对页面加载时刻（world.fixtureEpoch），按真实经过时间推进，不压缩。
 */
import type {
  JourneyLeg,
  JourneyMapSnapshot,
  LatLng,
  MapActivityEntry,
  PetArrivalContext,
  TransportMode,
  TransportNode,
  TravelActivity,
  TravellerRole,
  WorldService,
} from "@/shared/contracts";
import { atMin, atSec } from "./world";

const N = {
  hkg: { node_id: "fx-node-hkg", name: "香港机场（示意）", kind: "airport", timezone: "Asia/Hong_Kong", lat: 22.308, lng: 113.918, verified: false },
  hnd: { node_id: "fx-node-hnd", name: "东京羽田（示意）", kind: "airport", timezone: "Asia/Tokyo", lat: 35.549, lng: 139.779, verified: false },
  gzs: { node_id: "fx-node-gzs", name: "广州南站（示意）", kind: "station", timezone: "Asia/Shanghai", lat: 22.99, lng: 113.269, verified: false },
  wkl: { node_id: "fx-node-wkl", name: "香港西九龙（示意）", kind: "station", timezone: "Asia/Hong_Kong", lat: 22.304, lng: 114.166, verified: false },
  cen: { node_id: "fx-node-central", name: "中环码头（示意）", kind: "port", timezone: "Asia/Hong_Kong", lat: 22.287, lng: 114.157, verified: false },
  mfm: { node_id: "fx-node-macau", name: "澳门码头（示意）", kind: "port", timezone: "Asia/Macau", lat: 22.197, lng: 113.557, verified: false },
  hkc: { node_id: "fx-node-hk-city", name: "香港市区（示意）", kind: "road_point", timezone: "Asia/Hong_Kong", lat: 22.282, lng: 114.158, verified: false },
  szb: { node_id: "fx-node-sz-bay", name: "深圳湾（示意）", kind: "road_point", timezone: "Asia/Shanghai", lat: 22.515, lng: 113.944, verified: false },
} satisfies Record<string, TransportNode>;

const SERVICES: Record<string, WorldService> = {
  cat222: { world_service_id: "fx-ws-cat222", carrier_name: "喵航", service_code: "Cat222", mode: "flight", vehicle_style: "橘色机尾", reference_id: null, mapping_version: 1 },
  paw318: { world_service_id: "fx-ws-paw318", carrier_name: "爪爪铁路", service_code: "Paw318", mode: "train", vehicle_style: "奶白车厢", reference_id: null, mapping_version: 1 },
  otter08: { world_service_id: "fx-ws-otter08", carrier_name: "海獭轮渡", service_code: "Otter08", mode: "ferry", vehicle_style: "蓝白船身", reference_id: null, mapping_version: 1 },
};

function arc(a: LatLng, b: LatLng, bulge: number, steps = 16): LatLng[] {
  const points: LatLng[] = [];
  for (let i = 0; i <= steps; i += 1) {
    const t = i / steps;
    const lift = Math.sin(Math.PI * t) * bulge;
    points.push({ lat: a.lat + (b.lat - a.lat) * t + lift, lng: a.lng + (b.lng - a.lng) * t });
  }
  return points;
}

function ll(node: TransportNode): LatLng {
  return { lat: node.lat as number, lng: node.lng as number };
}

interface LegSeed {
  id: string;
  mode: TransportMode;
  role: TravellerRole;
  service: WorldService | null;
  from: TransportNode;
  to: TransportNode;
  departMin: number;
  arriveMin: number;
  route: LatLng[];
  schematic: boolean;
  arrived?: boolean;
}

function leg(seed: LegSeed): JourneyLeg {
  return {
    leg_id: seed.id,
    journey_id: "fx-journey-001",
    sequence: 1,
    kind: "main",
    mode: seed.mode,
    role: seed.role,
    world_service: seed.service,
    origin: seed.from,
    destination: seed.to,
    times: {
      origin_timezone: seed.from.timezone,
      destination_timezone: seed.to.timezone,
      planned_departure_utc: atMin(seed.departMin),
      planned_arrival_utc: atMin(seed.arriveMin),
      estimated_departure_utc: null,
      estimated_arrival_utc: null,
      actual_departure_utc: null,
      actual_arrival_utc: null,
    },
    time_basis: "demo_fixture",
    freshness: "unavailable",
    position_basis: seed.schematic ? "schematic" : "simulated_route",
    phase: seed.arrived ? "arrived" : "in_transit",
    itinerary_version: 1,
    route: seed.route,
    reference: { reference_id: null, time_basis: "demo_fixture", freshness: "unavailable", source_label: "演示行程，非真实班次", verified_at: null },
    rescheduled_reason: null,
  };
}

function activity(id: string, legId: string, kind: TravelActivity["kind"], state: TravelActivity["state"], startSec: number, endSec: number | null, sessionId: string | null): TravelActivity {
  return { activity_id: id, leg_id: legId, kind, state, starts_at: atSec(startSec), ends_at: endSec === null ? null : atSec(endSec), media_session_id: sessionId, interruptible: true, version: 1 };
}

function entry(a: TravelActivity, label: string, badge: "music" | "tv"): MapActivityEntry {
  return {
    entry_id: `entry-${a.activity_id}`,
    leg_id: a.leg_id,
    activity_id: a.activity_id,
    badge,
    badge_state: a.state,
    media_session_id: a.media_session_id,
    label,
    actions: a.state === "interrupted" ? ["solo", "open_leg_card"] : ["join", "solo", "open_leg_card"],
  };
}

function arrival(legSeed: LegSeed, city: string): PetArrivalContext {
  return {
    journey_id: "fx-journey-001",
    itinerary_version: 1,
    leg_id: legSeed.id,
    city,
    destination_timezone: legSeed.to.timezone,
    feasible_arrival_utc: atMin(legSeed.arriveMin + 20),
    stay_window_start_utc: atMin(legSeed.arriveMin + 30),
    stay_window_end_utc: atMin(legSeed.arriveMin + 150),
  };
}

export interface FixtureScenario {
  id: string;
  label: string;
  build: (nowMs: number) => JourneyMapSnapshot;
}

function snapshot(legSeed: LegSeed, activities: TravelActivity[], entries: MapActivityEntry[], arrivalCity: string | null, nowMs: number): JourneyMapSnapshot {
  const theLeg = leg(legSeed);
  // 活动是否仍在进行由服务器时间决定：结束后徽标消失（fixture 在此模拟服务端的活动窗口）。
  const liveActivities = activities.map((a) => (a.ends_at && Date.parse(a.ends_at) <= nowMs && a.state === "active" ? { ...a, state: "ended" as const } : a));
  const liveEntries = entries.filter((e) => liveActivities.find((a) => a.activity_id === e.activity_id)?.state !== "ended");
  return {
    journey_id: "fx-journey-001",
    pet_id: "fx-pet-001",
    itinerary_version: 1,
    lifecycle: "active",
    destination_title: arrivalCity ? `去${arrivalCity}（演示）` : "演示旅程",
    current_visit_id: null,
    planned_visit_id: "fx-visit-001",
    server_time: new Date(nowMs).toISOString(),
    legs: [theLeg],
    current_leg_id: theLeg.leg_id,
    vehicle: null,
    activities: liveActivities,
    activity_entries: liveEntries,
    arrival_context: arrivalCity ? arrival(legSeed, arrivalCity) : null,
    data_origin: "fixture",
  };
}

const flightSeed: LegSeed = { id: "fx-leg-flight", mode: "flight", role: "passenger", service: SERVICES.cat222, from: N.hkg, to: N.hnd, departMin: -60, arriveMin: 180, route: arc(ll(N.hkg), ll(N.hnd), 3.2), schematic: true };
const trainSeed: LegSeed = {
  id: "fx-leg-train",
  mode: "train",
  role: "passenger",
  service: SERVICES.paw318,
  from: N.gzs,
  to: N.wkl,
  departMin: -25,
  arriveMin: 22,
  route: [ll(N.gzs), { lat: 22.83, lng: 113.5 }, { lat: 22.6, lng: 113.86 }, { lat: 22.45, lng: 114.05 }, ll(N.wkl)],
  schematic: false,
};
const ferrySeed: LegSeed = { id: "fx-leg-ferry", mode: "ferry", role: "passenger", service: SERVICES.otter08, from: N.cen, to: N.mfm, departMin: -30, arriveMin: 30, route: [ll(N.cen), { lat: 22.26, lng: 113.98 }, { lat: 22.22, lng: 113.72 }, ll(N.mfm)], schematic: false };
const driveSeed: LegSeed = { id: "fx-leg-drive", mode: "drive", role: "driver", service: null, from: N.hkc, to: N.szb, departMin: -12, arriveMin: 40, route: [ll(N.hkc), { lat: 22.33, lng: 114.1 }, { lat: 22.42, lng: 114.03 }, { lat: 22.48, lng: 113.98 }, ll(N.szb)], schematic: false };

export const FIXTURE_SCENARIOS: FixtureScenario[] = [
  {
    id: "flight",
    label: "飞机·听歌",
    build: (now) => {
      // 听歌活动约 140 秒后结束 → 音符消失（验证“活动结束后更新图标”），之后是休息（无徽标）。
      const music = activity("fx-act-flight-music", flightSeed.id, "listening", "active", -20 * 60, 140, "fx-ms-flight");
      const rest = activity("fx-act-flight-rest", flightSeed.id, "resting", "active", 60, null, null);
      return snapshot(flightSeed, [music, rest], [entry(music, "TA 在听《窗边的小调》", "music")], null, now);
    },
  },
  {
    id: "train",
    label: "火车·看剧",
    build: (now) => {
      const video = activity("fx-act-train-video", trainSeed.id, "watching", "active", -5 * 60, 85, "fx-ms-train");
      return snapshot(trainSeed, [video], [entry(video, "TA 在看《云上色块》", "tv")], "香港", now);
    },
  },
  {
    id: "ferry",
    label: "轮船·无活动",
    build: (now) => snapshot(ferrySeed, [activity("fx-act-ferry-gaze", ferrySeed.id, "window_gazing", "active", -10 * 60, null, null)], [], "澳门", now),
  },
  {
    id: "drive",
    label: "驾车·只听",
    build: (now) => {
      const music = activity("fx-act-drive-music", driveSeed.id, "listening", "active", -5, null, "fx-ms-drive");
      return snapshot(driveSeed, [music], [entry(music, "TA 边开车边听歌", "music")], null, now);
    },
  },
  {
    id: "paused",
    label: "暂停",
    build: (now) => {
      const music = activity("fx-act-paused-music", flightSeed.id, "listening", "paused", -10 * 60, null, "fx-ms-paused");
      return snapshot(flightSeed, [music], [entry(music, "和 TA 一起暂停了", "music")], null, now);
    },
  },
  {
    id: "failed",
    label: "加载失败",
    build: (now) => {
      const video = activity("fx-act-failed-video", trainSeed.id, "watching", "active", -60, null, "fx-ms-failed");
      return snapshot(trainSeed, [video], [entry(video, "TA 在看纪录片", "tv")], "香港", now);
    },
  },
  {
    id: "arrived",
    label: "到站·已保存",
    build: (now) => {
      const arrivedSeed: LegSeed = { ...ferrySeed, id: "fx-leg-ferry-arrived", departMin: -70, arriveMin: -3, arrived: true };
      const video = activity("fx-act-arrived-video", arrivedSeed.id, "watching", "interrupted", -40 * 60, -3 * 60, "fx-ms-arrived");
      return snapshot(arrivedSeed, [video], [entry(video, "到站了，这集晚上接着看", "tv")], "澳门", now);
    },
  },
];

export const DEFAULT_SCENARIO = "flight";
